sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala

```diff
@@ -93,6 +93,21 @@ object SubqueryExpression {
     }.isDefined
   }
 
+  /**
+   * Returns true when an expression contains a correlated IN or correlated EXISTS and
+   * false otherwise. An IN is non-correlated only if the left values are foldable
+   * and the subquery has no outer references.
+   */
+  def hasCorrelatedInOrExists(e: Expression): Boolean = {
```

Comment:
a general question: what's the difference between this and `hasCorrelatedSubquery`? It seems to me that `In` or `Exists` are subqueries, so `hasCorrelatedSubquery` should cover it.

Reply:
@cloud-fan I believe @francis0407 does not want to consider the correlated scalar subquery here ? Also, he has an additional foldable check for LHS of in subq.

Reply:
Is my understanding corrected that `hasCorrelatedInOrExists` returns true if the input is `In` or `Exists` and `hasCorrelatedSubquery` returns true?

Reply:
Moreover, I think `a in (select b from c)` is a `Correlated In` with a `Non-Correlated Subquery`. And we are not going to use a subplan for it. We only need to deal with the `Non-Correlated In`, whose left value should be literal(or foldable) and the right subquery is non-correlated, such as `3 in (select b from c)`.

Reply:
@francis0407 I have a question. In this approach, are we going to get the result of the subquery to the driver ?

Reply:
@dilipbiswal Yes, currently we get all the result of the subquery to the driver. But we can use some methods (not in this PR) to reduce the result set in the executor side.

Reply:
@francis0407 Even if we push down the LHS of IN into subquery, there is no guarantee how much result we get back, right ? I think @cloud-fan needs to bless this.

Reply:
Push down the LHS of IN into the subquery is the first step. We can also use a specific aggregation to reduce the result set to an acceptable size.

Reply:
@francis0407 Great.. thanks.. when we do the aggregation, we can still outperform the existing left semi join approach, right ?

Reply:
Yes :)